```{r, include = FALSE}
ottrpal::set_knitr_image_path()
knitr::opts_knit$set(root.dir = '11-coronavirus')
# install packages required for module
install.packages("ape")
install.packages("harrietr")
```

# Coronavirus phylogenetics

In this lab, we'll use Nextstrain to explore the evolution of SARS-CoV-2 and then build our own coronavirus phylogenies within R.

#### Learning objectives

After completing this chapter, you'll be able to:

1. Interpret viral phylogenies and associated data on Nextstrain.
2. Explain how incomplete sampling affects phylogenies.
3. Describe how phylogenetic trees are built through neighbor joining.


## Phylogenetic trees

**Phylogenetic trees** represent the evolutionary relationships between samples. When comparing genetic sequences, we construct trees based on shared mutations between sequences -- with the caveat that sometimes the same mutation can arise independently on different lineages.

***
<details> <summary> Human vs. viral phylogenies </summary>

Constructing a phylogeny for human sequences is complicated by **recombination** (as well as diploidy and sex chromosomes), which means that every non-recombined segment of the genome has its own phylogenetic history.

Viral phylogenies are simpler to construct because they [recombine less frequently](https://www.nature.com/articles/nrmicro2614) than human chromosomes do. While the frequency of recombination varies by virus, [~3% of SARS-CoV-2 lineages](https://www.nature.com/articles/s41586-022-05189-9) show evidence of recombination (as opposed to 100% of human lineages, since recombination is an essential component of human meiosis).

Similar to the human Y chromosome, viral genomes typically accumulate mutations without recombination to split up haplotypes.

</details>
***


## Nextstrain

We'll be visualizing the evolutionary history of SARS-CoV-2 with a web application called [**Nextstrain**](https://nextstrain.org/ncov), developed by Trevor Bedford and colleagues at the Fred Hutch Cancer Center.

***
<details> <summary> Reading a phylogenetic tree </summary>

Nextstrain provides [**a guide**](https://docs.nextstrain.org/en/latest/learn/interpret/how-to-read-a-tree.html) on how to interpret their phylogenetic trees.

The figure below shows a "transmission tree" of viruses, where every dot is individual who was infected. Some of these individuals infect others, and some are dead ends (no transmission). After a mutation (diamonds) occurs on a branch, any sample to the right of the mutation will also carry it.

<center>

![**Fig. 1.** The "true" tree of virus transmission through a population ([source](https://docs.nextstrain.org/en/latest/learn/interpret/how-to-read-a-tree.html)).](11-coronavirus/images/infection_tree.png)

</center>

</details>
***


## Incomplete sampling

The tree we just saw is the complete, "full" tree of transmission, but we are almost never able to sample all the individuals we need to reconstruct it.

**Incomplete sampling** can result in inaccurate interpretations of a phylogenetic tree (ex: inferring fewer introductions of the virus from a source population into a new population).

<center>

![**Fig. 2.** Incomplete sampling complicates interpretations of a phylogeny ([source](https://docs.nextstrain.org/en/latest/learn/interpret/how-to-read-a-tree.html)).](11-coronavirus/images/tree_sampling.png)

</center>


## Tracking SARS-CoV-2 with phylogenetics

Go to [**the Nextstrain browser**](https://nextstrain.org/ncov/gisaid/global/all-time) to explore the phylogeny of SARS-CoV-2 sequences.

* Note that you can use the `PLAY` button on the left sidebar to watch its progression over time.
* Hovering over a specific node tells you the ID of the virus that was sequenced.
* It also tells you the **divergence**, which is the number of mutations it carries with respect to the root of the tree.

<center>

![**Fig. 3.** SARS-CoV-2 phylogeny from Nextstrain ([source](https://nextstrain.org/ncov/gisaid/global/all-time)).](11-coronavirus/images/phylogeny.png)

</center>

One example of the public health value of this phylogenetic analysis was the observation that the second reported case of novel coronavirus in Washington state (WA2) was genetically similar to the first reported case (WA1) six weeks earlier, suggesting ongoing community transmission. Click the tweet expand the full thread.

<blockquote class="twitter-tweet" data-lang="en"><p lang="en" dir="ltr"><a href="https://twitter.com/kevin_purcell">@kevin_purcell</a> Please feel free to try (click the menu Embed Tweet here) and let me know if it does not work.</p>&mdash; Yihui Xie (@xieyihui) <a href="https://twitter.com/trvrb/status/1233970442152472577">July 28, 2016</a></blockquote>
<script async src="//platform.twitter.com/widgets.js" charset="utf-8"></script>


## SARS-CoV-2 mutation landscape

The phylogeny page also has a "diversity" section, where it plots the number of mutations observed in different regions of the SARS-CoV-2 genome sequence.

<center>

![**Fig. 4.** Genomic distribution of mutations across SARS-CoV-2 strains ([source](https://nextstrain.org/ncov/gisaid/global/all-time)).](11-coronavirus/images/diversity.png)

</center>

***
<details> <summary> Does this plot imply that specific regions of the SARS-CoV-2 genome are more susceptible to mutation? </summary>

No -- this plot can be interpreted as showing where mutations **persist** in the SARS-CoV-2 genome, not where they _occur_.

Although mutation occurrence is mostly random, the genomic distribution we see in this plot has been filtered by natural selection. Beneficial mutations are selected for, causing them to appear more frequently in pathogenically important regions like the spike protein (**S**).

We sometimes see recurring independent mutations where the same amino acid is changed across different SARS-CoV-2 strains, which is extremely strong evidence of selection.

</details>
***


## Setup

In the rest of this module we'll place **SARS-CoV-2** on a phylogeny of 24 distantly related coronavirus sequences (SARS, MERS, etc.) from both human and animal hosts. 

### R packages

We'll use `ape` for phylogenetics, `ggtree` for plotting trees, and `harrietr` for data transformation.

```{r, message = FALSE, warning = FALSE}
library(tidyverse)
library(ape)
library(ggtree)
library(harrietr)
```


## Data

#### Metadata

The `accessions` dataframe contains the GenBank IDs and full names of the coronavirus sequences we're using:

```{r}
accessions <- read.table("accessions.txt", header = TRUE, sep = "\t")
head(accessions)
```

```{r}
# make vectors of the GenBank IDs and full names
# these will be used as input to functions later
ids <- accessions$id
names <- accessions$name
```

The SARS-CoV-2 sequence we're using is `MT093631` (`MT093631.2 Severe acute respiratory syndrome coronavirus 2 isolate SARS-CoV-2/human/CHN/WH-09/2020`).

#### DNA sequences

We've downloaded and **aligned** the genome sequences of these coronaviruses in the `aligned.fa` **FASTA** file. Click on the file to preview the sequence of the first coronavirus:

```
>DQ022305
----------------------------------------GTTAGGTTTTTACCTACCCAGGAAA--AGCCAACCAACC-
TTGATCTCTTGTAGATCTGTTCTCTAAACGAACTTTAAAA------TCTGTGTGGCTGTCGCTCGGCTGCATGCCTAGCG
CACCTACGCAGTATAAATATTAAT-AACTTTACTGTCGTTGACAAGAAACGAGTAACTCGTCCCTCTTCTGCAGACTGCT
```

***
<details> <summary> FASTA format </summary>

`.fa` indicates a **FASTA** file, which is a text-based format for representing DNA (or protein) sequences.

In a file that contains multiple sequences (like ours), the **`>`** character indicates the start of a new sequence and is usually followed by the sequence name.

</details>
***

***
<details> <summary> Why do the sequences have to be aligned? </summary>

To construct a phylogeny, we compare how a site in the genome has changed in different coronavirus strains. Sequences need to be **aligned** so that we know we're comparing the same site across sequences.

</details>
***


## Neighbor joining trees

One approach to building a phylogeny of sequences is **[neighbor joining](https://en.wikipedia.org/wiki/Neighbor_joining)**, which clusters sequences based on their pairwise genetic distance. In this approach, we:

* Start with a **star phylogeny** that assumes all samples are equally related (Fig. 5A)
* Compute a pairwise **distance matrix** between sequences, and look for the pair of sequences that are most similar to each other
* We join these two sequences to form a new node (Fig. 5B)
* The distance matrix is re-computed and this process repeats until all nodes are joined (Fig. 5C)

<center>

![**Fig. 5.** Steps for constructing a neighbor joining tree ([source](https://en.wikipedia.org/wiki/Neighbor_joining)).](11-coronavirus/images/neighbor_joining.jpg)

</center>


## Computing pairwise distance

Read in the FASTA file of aligned sequences with the `read.dna` function from `ape`:
 
```{r}
dna <- read.dna("aligned.fa", format = "fasta", as.matrix = TRUE)
```

We then compute the pairwise **distance matrix**:

```{r}
D <- dist.dna(dna, model = "TN93", as.matrix = TRUE)
```

We can plot this matrix to visualize it:

```{r}
# use the "melt_dist" function from harrietr package to convert
# the distance matrix to "long" format for ggplot
D_melted <- rbind(melt_dist(D, order = ids),
                  melt_dist(t(D), order = rev(ids)))

# plot distance matrix
ggplot(data = D_melted) +
  geom_tile(aes(x = iso1, y = iso2, fill = (dist + 1e-5))) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  scale_fill_viridis_c(name = "distance") +
  xlab("sample") +
  ylab("sample")
```

***
<details> <summary> Interpreting the distance matrix </summary>

Most of these coronaviruses seem to be fairly similar -- i.e., there's no clear clustering of sequences -- besides `HQ166910`, which looks genetically distinct from the other sequences.

</details>
***


## Building a phylogenetic tree

Using the distance matrix, we can now:

* Build a neighbor joining tree using the `nj()` function
* Use `HQ166910` as the [**outgroup**](https://en.wikipedia.org/wiki/Outgroup_(cladistics)) to root the tree (with the `root()` function)
* Use the `ladderize()` function to re-orient the tree into a tidier format for plotting

```{r}
# build a neighbor joining tree
tree <- nj(D)
# manually "root" the tree by setting HQ166910 as an outgroup
tree <- root(tree, which(ids == "HQ166910"))
# rotate tree at nodes to make it look tidier (i.e., "ladderized")
tree <- ladderize(tree)

# plot the tree
ggtree(tree) +
  theme_tree2() +
  geom_tiplab(label = names, size = 4) +
  xlim(0, 1.2)
```

On the tree, we can see that the 2019-nCov sample (`MT093631.2 Severe acute respiratory syndrome coronavirus 2`) groups most closely with `Bat coronavirus RaTG13`.

***
<details> <summary> Do you think this similarity is sufficient to confirm a bat origin of SARS-CoV-2? </summary>

Although the distance between SARS-CoV-2 and RaTG13 in the phylogeny looks small, it's a large distance in phylogenetic space. Without sampling more deeply within intermediate strains between RaTG13 and SARS-CoV-2, we don't know whether it passed through other mammalian species before being transmitted to humans.

</details>
***

yper
## Assess bootstrap support

A useful tool for evaluating confidence in a phylogenetic tree (or any other metric) is **[bootstrapping](https://en.wikipedia.org/wiki/Bootstrapping_\(statistics\))**. This statistical method is based on resampling data with replacement from the original dataset.

In our case, we resample aligned sites (i.e., bases) from the original alignment, then build a new tree with the resampled data. By repeating this procedure many times, we can evaluate confidence in various parts of the original tree by asking how often the trees from resampled data contain these features.

Run the code below to implement bootstrapping in the `boot.phylo()` function. The output is a vector of bootstrap support values, which we can overlay onto the tree.

```{r}
# set random seed
set.seed(123)
# bootstrap and build new trees to evaluate uncertainty
myBoots <- boot.phylo(tree, dna, 
                      function(x) ladderize(root(nj(dist.dna(x,
                                                             model = "TN93")),
                                                 which(ids == "HQ166910"))), 
                      rooted = TRUE)

# replace "NA" with zero in bootstrap results; do not label terminal nodes
myBoots[is.na(myBoots)] <- 0
myBoots <- c(rep(NA, 25), myBoots)

# re-plot tree with bootstrap values
ggtree(tree, branch.length = "none") +
  theme_tree2() +
  geom_tiplab(label = names) +
  geom_label(aes(label = myBoots), size = 3) +
  xlim(0, 15)
```


## Conclusion

In this module, we visualized evolutionary relationships between coronavirus strains using Nextstrain and by constructing our own phylogenies in R.

* We first explored [**Nextstrain**](https://nextstrain.org/ncov/gisaid/global/all-time), a browser that summarizes historical sequencing data from SARS-CoV-2 strains around the world.
<br></br>
* We observed independent **recurrent mutations** in pathogenically relevant regions of the SARS-CoV-2 genome, suggesting that such mutations confer a fitness benefit and are under positive selection.
<br></br>
* Using the sequences of other coronavirus strains from GenBank, we placed SARS-CoV-2 onto a **neighbor joining** phylogeny of closely related viruses.
  * We used **bootstrapping** to provide confidence estimates for branches on our neighbor joining tree.
  

## Homework

#### Learning Objectives

* Practice building and interpreting phylogenies in R

#### Assignment

`hw_accessions.txt` and `hw_aligned.fa` provide the alignments and metadata for 24 human SARS-CoV-2 sequences, plus the bat RaTG13 coronavirus strain.

Use this data to repeat the analysis we did in class (compute distance matrix, construct neighbor joining tree, perform bootstrapping for confidence values) to build a phylogeny of SARS-CoV-2 strains. You should use RaTG13 (accession: `MN996532`) as an outgroup.

***
<details> <summary> Solution </summary>

Read in accession data:

```{r}
accessions <- read.table("hw_accessions.txt", sep = "\t", header = TRUE)
# make vectors of accessions IDs and full names
ids <- accessions$id
names <- accessions$name
```

Construct distance matrix:

```{r}
# read fasta file
dna <- read.dna("hw_aligned.fa", format = "fasta", as.matrix = TRUE)
# compute pairwise distance matrix
D <- dist.dna(dna, model = "TN93", as.matrix = TRUE)
```

Build neighbor joining tree:

```{r}
# make the tree
tree <- nj(D)
# manually root tree at RaTG13 coronavirus
tree <- root(tree, which(ids == "MN996532"))
# ladderize tree
tree <- ladderize(tree)
```

Bootstrap to determine tree uncertainty:

```{r}
set.seed(123)
myBoots <- boot.phylo(tree, dna, 
                      function(x) ladderize(root(nj(dist.dna(x,
                                                             model = "TN93")),
                                                 which(ids == "MN996532"))), 
                      rooted = TRUE)
# replace "NA" with zero
myBoots[is.na(myBoots)] <- 0
```

Plot the bootstrapped tree:

```{r}
ggtree(tree, branch.length = "none") +
  theme_tree2() +
  geom_tiplab(label = names) +
  geom_nodelab(label = myBoots, geom = "label", fill = "#deebf7") +
  xlim(0, 25)
```

</details>
***