```{r, include = FALSE}
ottrpal::set_knitr_image_path()
knitr::opts_knit$set(root.dir = '09-archaic_admixture')
# install packages for the lab
install.packages("admixr")
```

# Archaic admixture

In this lab, we'll discuss three common statistics, $D$, $f_{4}$, and the $f_{4}$-ratio, which are used for quantifying and testing hypotheses related to admixture (or "introgression").

#### Learning objectives

After completing this chapter, you'll be able to:

1. Explain how incomplete lineage sorting can create patterns of allele sharing that are discordant with species relationships.
2. Describe how introgression is expected to affect patterns of allele sharing.
3. Interpret the $D$ and $f_4$ statistics and their associated p-values.
4. Interpret the $f_{4}$-ratio ratio statistic and whether it implies differences in archaic admixture between populations.


## Neanderthal and Denisovan introgression

As modern humans migrated out of Africa 50,000-70,000 years ago, they encountered and intebred with two groups of **archaic hominins**, the Neanderthals and Denisovans.

Admixture with archaic hominins resulted in **introgression** of archaic DNA into these migrating populations. These Neanderthal and Denisovan sequences still persist in the genomes of their present-day descendents.

<center>

![**Fig. 1.** Ancient encounters between migrating human populations and archaic hominins.](08-archaic_admixture/images/migrations.jpg)

</center>


## Inferring introgression from phylogenetic trees

One test for introgression -- called the $\textbf{D}$ **statistic**, or **ABBA-BABA test** -- compares the phylogenetic tree of species-level relationships with the tree for a specific genetic variant.

We expect the **species-level tree** for Neanderthals and humans to look like the white bars of the tree in **Fig. 2**, where any two human populations are more similar to each other than they are to Neanderthals or chimpanzees (the outgroup).

If a genetic variant differs between humans and Neanderthals, we generally expect it to segregate according to the species-level tree, in an `AABB` pattern (**Fig. 2**).

<center>

![**Fig. 2.** A variant concordant with the species-level tree for modern humans and Neanderthals.](08-archaic_admixture/images/aabb.jpg){width=60%}

</center>


## Incomplete lineage sorting

More rarely, the tree constructed for a specific SNP is discordant with the species tree. This is usually due to **incomplete lineage sorting (ILS)** -- when genetic variation is present in the ancestor of two populations (e.g., the modern human--Neanderthal ancestor), then randomly fixes in a fashion that happens to conflict with the populations' evolutionary relationships.

ILS can create discordant **`ABBA`** or **`BABA`** trees for a given SNP. Under ILS, we would expect to see approximately equal numbers of `ABBA` and `BABA` trees.

<center>

![**Fig. 3.** ABBA and BABA allele patterns can be formed by incomplete lineage sorting.](08-archaic_admixture/images/ils.jpg)

</center>


## Evidence of introgression

Discordant trees can also be produced by **introgression**. In this case, the B allele arises on the Neanderthal lineage and is passed from Neanderthals into a human population, creating **either an `ABBA` or `BABA` pattern** (depending on which population receives Neanderthal gene flow).

<center>

![**Fig. 4.** ABBA and BABA allele patterns formed by Neanderthal introgression.](08-archaic_admixture/images/introgression.jpg)

</center>


## The $D$ statistic

Introgression creates an imbalance in the number of `ABBA` or `BABA` trees if only one of the human populations has admixed with Neanderthals. The $\textbf{D}$ **statistic** quantifies this imbalance:

$$
D = \frac{\textrm{# BABA sites} - \textrm{# ABBA sites}}{\textrm{# BABA sites} + \textrm{# ABBA sites}}
$$

$D > 0$ is evidence for Neanderthal gene flow into the H2 population, while $D < 0$ is evidence for gene flow into H1.

***
<details> <summary> Choice of populations for the $D$ statistic </summary>

The choice of populations is very important when calculating the $D$ statistic.

When assessing archaic introgression, **H2** is typicaly set as a human population without archaic admixture (e.g., a population from Africa). If H2 instead were a European population that did possess introgressed sequence, we would not expect a significant $D$ statistic.

</details>
***


## Setup

In this lab, we'll quantify evidence for introgression in human populations using $D$ and related statistics.

### R packages

In addition to `tidyverse`, we'll use the `admixr` package, which allows us to easily run the software package [ADMIXTOOLS](https://github.com/DReichLab/AdmixTools) from within R to calculate $D$ and related statistics.

The exercises for class were adapted from the `admixr` tutorial, available [here](https://bodkan.net/admixr/articles/01-tutorial.html).

```{r, results = FALSE, message = FALSE, warning = FALSE}
library(tidyverse)
library(admixr)
```


## Data

The `admixr` package provides real example data from 10 human individuals, which can be acquired by running its `download_data()` function:

```{r, eval = FALSE}
# download data into current directory
prefix <- download_data(dirname = ".")
```

We now have a directory called `snps` that contain four files:

* `snps.geno`: Genotype of each individual (column) at each SNP (row)
  * Represented as counts of the alternative allele (`0`, `1`, `2`)
* `snps.ind`: Population IDs for each individual
* `snps.snp`: SNP IDs, positions, and alleles
* `regions.bed`: A file of genomic regions (not required for basic `admixr` analysis)

***
<details> <summary> EIGENSTRAT format </summary>

Together, the three `.geno`, `.ind`, and `.snp` files constitute **EIGENSTRAT format**. This is just a way of representing genotype data, similar to a VCF -- in fact, several software packages exist to convert between VCF and EIGENSTRAT.

</details>
***


## Reading in data

We can provide the location of the downloaded files to the `eigenstrat()` function, which then constructs an EIGENSTRAT object to be used for downstream analysis.

```{r, eval = FALSE}
# read in eigenstrat files
snps <- eigenstrat("./snps/snps")
snps
```
```
## EIGENSTRAT object
## =================
## components:
##   ind file: ./snps/snps.ind
##   snp file: ./snps/snps.snp
##   geno file: ./snps/snps.geno
```

## The `d()` function

`admixr` computes the $D$ statistic in a function called `d()`.

```{r}
?d
```

As input, `d()` asks for:

* Data object in EIGENSTRAT format
* Four population names (`W`, `X`, `Y`, `Z`) to calculate D between


## Computing the D statistic

Let's compute $D$ for four of the individuals we have data for: French, Sardinian, Vindija (Neanderthal), and chimpanzee.

```{r, eval = FALSE}
d_result <- d(data = snps,
              # provide population names to calculate D between
              W = "French", X = "Sardinian", Y = "Vindija", Z = "Chimp")

d_result
```
```{r read-snps-data, echo = FALSE}
# read in table with the result data bc we can't run admixr on bookdown
d_result <- read.table("https://drive.google.com/uc?export=download&id=10eHG6hbuSxxUTXrWMc9ubBnSvuIhQQ6F", header = TRUE)
d_result
```


***
<details> <summary> How do we interpret these results? </summary>

The last three columns count the number of **ABBA** and **BABA** sites, as well as the total number of variants being analyzed. First, note that the ABBA/BABA sites are only a small fraction of the total number of variants -- most variants conform to the species-level tree.

The number of ABBA and BABA variants also looks similar, which implies that the discordant trees in these four populations primarily result from **ILS** rather than introgression.

The middle columns give the actual value of $D$ and its standard error, as well as the Z score (which is equal to $\frac{D}{\textrm{stderr}}$).

</details>
***


## Converting to p-values

How do we know whether the ABBA-BABA counts are significantly different? We can convert the Z score into a p-value:

```{r}
d_result <- d_result %>%
  # convert z score into pvalue
  mutate(p = 2 * pnorm(-abs(Zscore)))

d_result
```

***
<details> <summary> Interpretation of this p-value </summary>

The p-value does not look significant (no introgression into the ancestors of this French or Sardinian individual). This is in line with our observation that there doesn't seem to be a disproportionate amount of ABBA or BABA sites.

</details>
***


## Computing D for all populations

We can give `d()` a list of populations to compute the $D$ statistic for, where:

* `W`: A vector of our populations of interest
* `X`: An assumed **unadmixed** population (Yoruba)
* `Y`: The **introgressing** population (Vindija Neanderthal)
* `Z`: Outgroup (chimpanzee)

```{r, eval = FALSE}
# create vector of populations of interest
pops <- c("French", "Sardinian", "Han", "Papuan",
          # including three African populations
          "Khomani_San", "Mbuti", "Dinka")

# calculate D
d_result <- d(data = snps,
              W = pops, X = "Yoruba", Y = "Vindija", Z = "Chimp") %>%
  # convert z score into pvalue
  mutate(p = 2 * pnorm(-abs(Zscore)))

d_result
```
```{r, echo = FALSE}
# read in table with the result data bc we can't run admixr on bookdown
d_result <- read.table("https://drive.google.com/uc?export=download&id=1U35MiovpfdChwLgBzEW8dEjr6NHExa1b", header = TRUE)
d_result
```


***
<details> <summary> Interpretation of $D$ results </summary>

This is the result that was published in [**Green et al. 2010**](https://www.science.org/doi/full/10.1126/science.1188021), revealing evidence for gene flow from Neanderthals into the ancestors of non-African populations.

We see significant evidence of Neanderthal introgression into the genomes of the non-African samples (French, Sardinian, Han, Papuan), but not the African samples (although we know from recent research that this is an oversimplification).

</details>
***


## Plotting the D statistic

Use this code to plot the $D$ statistic and standard error calculated for each population. This is a visual representation of the data in the `d_result` table:

```{r}
ggplot(data = d_result,
       aes(x = W, y = D, color = p < 0.05)) +
  geom_point() +
  geom_hline(yintercept = 0, linetype = 2) +
  geom_errorbar(aes(ymin = D - 1.96 * stderr, ymax = D + 1.96 * stderr),
                width = 0.5) +
  xlab("Population")
```


## $f_{4}$ statistic

The $\mathbf{f_{4}}$ **statistic** -- not to be confused with the $\mathrm{F_{ST}}$ from the previous week -- is very similar to the *D* statistic. Its main advantage is that it is proportional to the branch length separating two pairs of populations.

Compute the $f_{4}$ statistic for all populations using the code below:

```{r, eval = FALSE}
f4_result <- f4(data = snps,
                W = pops, X = "Yoruba", Y = "Vindija", Z = "Chimp") %>%
  # convert z score into pvalue
  mutate(p = 2 * pnorm(-abs(Zscore)))

f4_result
```
```{r, echo = FALSE}
# read in table with the result data bc we can't run admixr on bookdown
f4_result <- read.table("https://drive.google.com/uc?export=download&id=1A6xOGWwV7k3PsUP9RLIo1j1eswFX9Liw", header = TRUE)
f4_result
```

Note that the p-values are the same as when we calculated the $D$ statistic, but the actual $f_4$ values are different.


## $f_{4}$-ratio statistic

The branch length proportionality of the $f_4$ statistic is useful for deriving yet another statistic, called the **$\mathbf{f_{4}}$-ratio statistic**.

As implied by its name, this simply a ratio of two different $f_{4}$ statistics. Unlike $D$ and $f_{4}$, the $f_{4}$-ratio tells us _how much_ Neanderthal ancestry a given individual possesses.

Calculate the $f_{4}$-ratio using the code block below:

```{r, eval = FALSE}
f4_ratio_result <- f4ratio(data = snps,
                           X = pops, A = "Altai", B = "Vindija", C = "Yoruba", O = "Chimp") %>%
  # convert z score to pvalue
  mutate(p = 2 * pnorm(-abs(Zscore)))

f4_ratio_result
```
```{r, echo = FALSE}
# read in table with the result data bc we can't run admixr on bookdown
f4_ratio_result <- read.table("https://drive.google.com/uc?export=download&id=1vrzYZvSsZuIHgssMJ7SiDvjG1RLycuLe", header = TRUE)
f4_ratio_result
```

For this statistic, `alpha` represents the proportion of the genome whose ancestry traces to Neanderthal introgression.


## Plotting $f_{4}$-ratio results

Run the code below to plot the $f_{4}$-ratio values we computed:

```{r}
ggplot(data = f4_ratio_result,
       aes(x = X, y = alpha, color = p < 0.05)) +
  geom_point() +
  geom_errorbar(aes(ymin = alpha - 2 * stderr, ymax = alpha + 2 * stderr),
                width = 0.5) +
  geom_hline(yintercept = 0, linetype = 2) +
  labs(y = "Neanderthal ancestry proportion", x = "Present-day individual")
```

***
<details> <summary> Based on what we know about Papuan populations, do you think the estimate of 4% Neanderthal introgressed DNA is accurate? </summary>

We know that Oceanian populations also experienced introgression from Denisovans. Denisovans were more genetically similar to Neanderthals than to modern humans, so Denisovan ancestry is being counted as Neanderthal ancestry by this metric.

</details>
***


## Computing statistics in genomic intervals

We've so far computed the $D$, $f_4$, and $f_4$-ratio statistics across the entire genome, but we can also restrict computation to particular genomic intervals, such as genes, promoters, enhancers, etc.

***
<details> <summary> What would region-specific differences imply? </summary>

Region-specific differences suggest that there are some regions of the genome that are particularly tolerant or intolerant to introgression.

</details>
***

<center>

![**Fig. 5 ([source](https://www.pnas.org/doi/full/10.1073/pnas.1814338116)).** Proportion of introgressed Neanderthal ancestry within specific gene annotations.](08-archaic_admixture/images/petr.jpeg){width=85%}

</center>

In the figure below, we see that the genome-wide average percentage of Neanderthal ancestry is ~2%, but there are some genomic regions (promoters, evolutionarily conserved sequences) where this percentage is lower.

One theory explaining this pattern is that Neanderthals accumulated a genetic load of slightly deleterious mutations due to their small population sizes. Consequently, when Neanderthal sequences were introgressed into human populations, they were subjected to negative selection, which was stronger in more functionally important regions of the genome.


## BED files

Annotations of **genomic intervals** are commonly represented in a file format called [**BED**](https://genome.ucsc.edu/FAQ/FAQformat.html#format1).

We can utilize these files in `admixr`'s `filter_bed()` function, which excludes or restricts analyses to genomic intervals within a BED file.

Run the code below to re-calculate the $f_4$-ratio with `promoters_hg19.bed` -- a bed file containing the coordiinates for promoters annotated in the [Ensembl Regulatory Build](https://useast.ensembl.org/info/genome/funcgen/data/accessing-regulation.html). We can create a new `snps` data object that either **keeps** or **excludes** these regions from our analysis.

```{r, eval = FALSE}
# get the path to the `regions.bed` file
bed <- file.path("promoters_hg19.bed")

# option 1: KEEP only these regions for analysis
new_snps_keep <- filter_bed(snps, bed)
# option 2: REMOVE these regions from analysis
new_snps_remove <- filter_bed(snps, bed, remove = TRUE)
```


## Region-specific $f_4$ ratio

Now we can re-calculate the $f_4$-ratio only within promoter regions.

```{r, eval = FALSE}
# f4-ratio with only promoters
f4_filtered <- f4ratio(data = new_snps_keep,
                       X = pops, A = "Altai", B = "Vindija", C = "Yoruba", O = "Chimp") %>%
  # convert z score to pvalue
  mutate(p = 2 * pnorm(-abs(Zscore)))

f4_filtered
```
```{r, echo = FALSE}
# read in table with the result data bc we can't run admixr on bookdown
f4_filtered <- read.table("https://drive.google.com/uc?export=download&id=15sqNkRpzhe4WbuBAVyj-XQNuB33DF2hg", header = TRUE)
f4_filtered
```

***
<details> <summary> Plot the region-excluded $f_4$-ratios </summary>

```{r}
ggplot(f4_filtered,
       aes(x = X, y = alpha, color = p < 0.05)) +
  geom_point() +
  geom_errorbar(aes(ymin = alpha - 2 * stderr, ymax = alpha + 2 * stderr), width = 0.5) +
  geom_hline(yintercept = 0, linetype = 2) +
  labs(y = "Neanderthal ancestry proportion", x = "Present-day individual")
```

Except for the Han population, we see almost no Neanderthal ancestry when we calculate the $f_4$-ratio within promoters -- supporting the idea that [functionally important genomic regions are depleted for Neanderthal introgression](https://www.pnas.org/doi/full/10.1073/pnas.1814338116).

</details>
***


## Conclusion

In this lab, we assessed evidence for Neanderthal introgression into specific human populations.

* We reviewed the $\mathbf{D}$ **statistic**, which asks whether a variant shows an imbalance in the population inheritance patterns expected from **incomplete lineage sorting (ILS)**. Such an imbalance implies historical introgression into one population.
<br></br>
* With `admixr`, we calculated the $D$ statistic for one individual from each of seven modern human populations, identifying significant evidence for Neanderthal introgression into the ancestors of non-African -- but not African -- populations.
<br></br>
* We also calculated the $\mathbf{f_4}$ **statistic**, which is very similar to the $D$ statistic but is proportional to the genetic distance between two populations.
<br></br>
* We computed the $\mathbf{f_4}$**-ratio**, whose value indicates the proportion of introgressed ancestry in each population.
* Finally, we computed the $\mathbf{f_4}$**-ratio** within just promoter regions, where we observed a depletion of Neanderthal ancestry.


## Homework

#### Learning Objectives

* Practice calculating introgression statistics in `admixr`
* Interpret the biological significance of region-specific values of the $f_4$-ratio

#### Assignment

Follow these steps to create your own genome stratifications for calculating the $f_4$ ratio statistic.

1. Go to the [**UCSC Table Browser**](http://genome.ucsc.edu/cgi-bin/hgTables), where you can find a wide selection of annotations for the human genome.
2. Make sure you set the `assembly:` drop-down box to `Feb. 2009 (GRCh37/hg19)`.
3. Use the `group:` and `track:` menus to select any set of genomic regions.
    * You can click the `data format description` button and scroll to the `Description` section to find out what each annotation represents.
4. Under the `Retrieve and display data` section, set the `output format:` to BED.
5. Enter an `output filename:` (ex: `all_genes.bed`).
6. Click `get output` to download the file.
7. In Posit Cloud, upload your file using the `Upload` button in the `Files` panel (bottom right).
8. **Run the code block below** to reformat the BED file. The code matches the UCSC's chromosome naming format with the format used in the `snps` data:

```{r, eval = FALSE}
# fill in blank with the name of your bed file
system(command = "sed -i 's/chr//g' ________")

# get the path to your bed file
bed <- file.path("________")
```

Compute the $f_4$ ratio statistic within and outside of the genomic intervals. Repeat for another set of genome annotations to contrast Neanderthal ancestry in different genomic elements.

***
<details> <summary> Solution </summary>

Download tracklist of haploinsufficient genes (`Phenotype and Literature` -> `Haploinsufficiency`).

```{r, eval = FALSE}
# get the path to the `regions.bed` file
bed <- file.path("haploinsufficient.bed")

# option 1: KEEP only these regions for analysis
new_snps_keep <- filter_bed(snps, bed)
# option 2: REMOVE these regions from analysis
new_snps_remove <- filter_bed(snps, bed, remove = TRUE)
```

Re-calculate the $f_4$-ratio:

```{r, eval = FALSE}
# f4-ratio with the regions kept
f4_keep <- f4ratio(data = new_snps_keep,
                       X = pops, A = "Altai", B = "Vindija", C = "Yoruba", O = "Chimp") %>%
  # convert z score to pvalue
  mutate(p = 2 * pnorm(-abs(Zscore)))

f4_keep
```
```{r, echo = FALSE}
# read in table with the result data bc we can't run admixr on bookdown
f4_keep <- read.table("https://drive.google.com/uc?export=download&id=15fCANqewJ5du3gzZLIc1Kv2C1yz4s4DQ", header = TRUE)
f4_keep
```

```{r, eval = FALSE}
# f4-ratio with the regions removed
f4_remove <- f4ratio(data = new_snps_remove,
                       X = pops, A = "Altai", B = "Vindija", C = "Yoruba", O = "Chimp") %>%
  # convert z score to pvalue
  mutate(p = 2 * pnorm(-abs(Zscore)))

f4_remove
```
```{r, echo = FALSE}
# read in table with the result data bc we can't run admixr on bookdown
f4_remove <- read.table("https://drive.google.com/uc?export=download&id=1ZrVRsY4A3VpaaD9PUbSmkJX_1coJeTen", header = TRUE)
f4_remove
```

Some of the alpha values for each population change when excluding/restricting to haploinsufficient genes, but their standard error ranges still overlap between the two $f_4$-ratio calculations, so they likely aren't truly different.

</details>
***
